devilspie2.1
.\"
.\" Copyright © 2011-2014 Andreas Rönnquist.
.\" This file is distributed under the same license
.\" as the devilspie2 package, see COPYING file.
.\"
.TH devilspie2 1 "" ""
.SH NAME
Devilspie2 - perform scripted actions on windows as they are created
.SH Description
.B Devilspie2
is a program that detects windows as they are created, and performs scripted
actions on them. The scripts are written in LUA, allowing a great deal of
customisation.

More details in /usr/share/doc/devilspie2.

.SH Options
.IP "--version"
Displays the version of Devilspie2.
.IP "--folder folder"
Sets the folder where the LUA scripts are loaded from.
.IP "--debug"
Shows debug information from the LUA scripts. If debug_print is used in the LUA
scripts, that output will only be printed to stdout if this option is used.
.IP "--emulate"
Emulate script behaviour. Using this option only emulates a scripts behaviour,
and doesn't affect the windows that the scripts otherwise would modify.
.IP "--wnck-version"
Show version of the libwnck used. (Only availible on GTK3 or later)
.SH Script Commands

.TP
See the README or the manual linked from the homepage for a detailed description
of the commands recognized by the LUA script.

.SH Author
.B Devilspie2
was written by Andreas Rönnquist <devilspie2@gusnan.se>, and is based on Devilspie
by Ross Burton.


ChangeLog

ChangeLog:
----------

0.36
	* _Really_ fix the man page commenting style
	* Remove version and date from man-page

0.35
	* Fix man page commenting

0.34
	* Add GPL headers in more places where needed
	* Fix cppcheck reassignment warning
	* Install man-page with Makefile too

0.33
	* Add boolean argument to set_window_{above,below} (Thanks to rbtwrrr)
	* Fix spelling of horizontally in all places
	* Various source formatting fixes

0.32
	* Add functions get_fullscreen (get_window_fullscreen)
	* Small documentation fixes
	* Add set_window_strut function
	* gdk_display_get_n_screens is deprecated on GLIB >= 3.10, there we default
	  to only one screen.
	* Updated Russian translation

0.31
	* Add close_window function
	* Remove unused C function my_wnck_get_cardinal
	* Fix string problems reported by pscan
	* Updated Russian translation
	* Added Japanese translation

0.30
	* Update French translation
	* Add function set_window_opacity to do the same as set_opacity
	* Add maximize_horizontally with same functionality as
	  maximize_horisontally
	* Add get_window_is_maximized_horizontally with same functionality as
	  get_window_is_maximized_horisontally
	* Add focus and focus_window functions
	* Updated Brasilian-Portugese translation

0.29
	* Add finnish translation
	* Updated dutch translation
	* Add examples to get_window_geometry and get_window_client_geometry
	* Add get_screen_geometry function (Thanks to Ronan Barzic)

0.28
	* Make get workspace error string consistent with other past-tense strings
	* wnck_class_group_get_res_class is deprecated, use wnck_class_group_get_id
	  instead on wnck versions it isn't available
	* Make 'relative window positioning' possible, allowing negative values to
	  set_window_geometry
	* Remove sciteproj project file
	* Add set_window_type function
	* Add set_window_opacity function
	* Added handling of window closing events
	* Add file monitor to automatically load added scripts

0.27
	* Update Makefile to fix a LUA problem with pkg-config
	* Update the email contact address in the README
	* Add get_workspace_count function - Thanks Hans Nieser!
	* Use luaL_newstate instead of lua_open which is deprecated

0.26
	* Add "center" function
	* Register set_window_position2 under the correct name - Fixes
	  bug where set_window_position2 couldn't be found
	* Add a text file describing development for people who
	  want to join
	* Add a TODO file
	* Updated translations

0.25
	* Add set_viewport function
	* Move error strings to a source file of their own
	* Updated translations - Thanks guys!
	* Make it possible to build using -j without screwing up
	  translation build

0.24
	* Fix bug where folder settings got overridden
	  - Thanks Andreas Müller!
	* Add set_window_fullscreen function
	* Add some required checks for running in emulation mode

0.23
	* Added Russian translation - Thanks to Roman V. Nikolaev
	* Fix a string memory leak
	* Some source cleanups

0.22
	* Make scripts keep their environment between runs - Makes it possible
	  to do something to a window class the "n-th" time a window of that
	  class opens.
	* Documentation and translation updates.

0.21
	* Add get_window_is_maximized, get_window_is_maximized_horisontally
	  and get_window_is_maximized_vertically functions
	* Add set_window_geometry2 and set_window_position2 functions,
	  giving more similar behaviour to original devilspie functionality
	* Add set_window_above and set_window_below functions
	* Add get_window_type function
	* Add get_class_instance_name function
	* Add get_window_property function
	* Add get_window_role function
	* Add get_window_xid function
	* Add get_window_class function
	* Sort the list of scripts loaded, making it easier to determine script
	  loading order.
	* Require an X11 display to run on GTK3
	* Updated translations (Thanks guys!)

0.20
	* Add get_window_has_name function
	* Added Brasilian Portugese translation
	  - thanks to Hélder Máximo Botter Ribas
	* Massive whitespace fixes in the source
	* Documentation updates
	* Add linking to X11 (fixes linker error on some ports)

0.19
	* Add C preprocessor flags to build system
	* Minor update to documentation for translators

0.18
	* Translation fixes
	* Minor translation documentation fixes
	* Minor cleanups and fixes
	* Added French translation - thanks to Robin Hahling

0.17
	* Fix build warning for string not using %s

0.16
	* Documentation updates
	* Add get_window_geometry and get_client_window_geometry functions
	* Add option to get system wnck version (on systems that support it)
	* Added gettext support for translations, and added Swedish translation
	* Add set_skip_tasklist function
	* Add set_skip_pager function

0.15
	* Fix GPL headers, which incorrectly was LGPL
	* Minor documentation fixes

0.14
	* Better checks for valid windows
	* Source code commenting updates
	* Minor documentation updates

0.13
	* Updated example in the README and other documentation updates
	* Makefile updates

0.12
	* Fixes to the Makefile, fixing the depends system, fixing installation
	  prefxes

0.11
	* Fix a build problem where standard C-flags wouldn't be included in makefile
	* Fix incorrect debug stuff to the makefile

0.10
	* Fixed a problem that made the project link even if no objects had changed
	* Default to build using GTK3, and add option to build using GTK2
	* Documentation fixes

0.09
	* Fixes to the Makefile
	* Minor documentation updates

0.08
	* Fix manpage problems

0.07
	* Fixes to the build system
	* Fixes of previous sloppy releases
	* Minor documentation updates

0.05
	* Update Makefile to allow CFLAGS changes, required by CDBS

0.04
	* Build system updates

0.03
	* Remove CMake stuff and go to a standard Makefile
	* Documentation updates
	* Move installation instructions to file of its own

0.02
	* Rename set_workspace to set_window_workspace
	* Massive documentation updates

0.01
	* Initial release


--------------------------------------------------------------------------------


src/devilspie2.c
/**
 *	This file is part of devilspie2
 *	Copyright (C) 2005 Ross Burton, 2011-2013 Andreas Rönnquist
 *
 *	devilspie2 is free software: you can redistribute it and/or
 *	modify it under the terms of the GNU General Public License as published
 *	by the Free Software Foundation, either version 3 of the License, or
 *	(at your option) any later version.
 *
 *	devilspie2 is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *	GNU General Public License for more details.
 *
 *	You should have received a copy of the GNU General Public License
 *	along with devilspie2.
 *	If not, see <http://www.gnu.org/licenses/>.
 */

#include <string.h>

#include <stdlib.h>
#include <glib.h>
#include <glib/gstdio.h>
#include <gdk/gdk.h>
#include <gdk/gdkx.h>
#include <glib/gi18n.h>

#define WNCK_I_KNOW_THIS_IS_UNSTABLE
#include <libwnck/libwnck.h>

#include <lua.h>
#include <lualib.h>
#include <lauxlib.h>

#include <locale.h>

#include "script.h"
#include "script_functions.h"

#include "error_strings.h"

#include "config.h"


/**
 *
 */
GMainLoop *loop = NULL;

static gboolean debug = FALSE;
static gboolean emulate = FALSE;
static gboolean show_version = FALSE;

#ifdef HAVE_GTK3
	// libwnck Version Information is only availible if you have
	// libwnck 3.0 or later
	static gboolean show_wnck_version = FALSE;
#endif

static gchar *script_folder = NULL;
static gchar *temp_folder = NULL;

GFileMonitor *mon = NULL;

gchar *config_filename = NULL;

/**
 *
 */
static void load_list_of_scripts(WnckScreen *screen, WnckWindow *window,
											GSList *file_list)
{
	GSList *temp_file_list = file_list;
	// set the window to work on
	set_current_window(window);

	// for every file in the folder - load the script
	if (file_window_open_list != NULL) {

		while(temp_file_list) {
			gchar *filename = (gchar*)temp_file_list->data;

			// is it a LUA file?
			if (g_str_has_suffix((gchar*)filename,".lua")) {

				// init the script, run it
				if (!load_script(global_lua_state, filename)) {
				}

				run_script(global_lua_state);

			}
			temp_file_list=temp_file_list->next;
		}
	}
	return;

}


/**
 *
 */
static void window_opened_cb(WnckScreen *screen, WnckWindow *window)
{
	load_list_of_scripts(screen, window, file_window_open_list);
}


/**
 *
 */
static void window_closed_cb(WnckScreen *screen, WnckWindow *window)
{
	load_list_of_scripts(screen, window, file_window_close_list);
}


/**
 *
 */
void init_screens()
{
	int i;
	int num_screens;

#ifndef GDK_VERSION_3_10
	num_screens = gdk_display_get_n_screens(gdk_display_get_default());
#else
	num_screens = 1;
#endif

	for (i=0; i<num_screens; i++) {
		WnckScreen *screen = wnck_screen_get(i);

		g_signal_connect(screen, "window-opened",
								(GCallback)window_opened_cb, NULL);
		g_signal_connect(screen, "window-closed",
								(GCallback)window_closed_cb, NULL);
	}
}


/**
 * atexit handler - kill the script
 */
void devilspie_exit()
{
	clear_file_lists();

	if (temp_folder != NULL) g_free(temp_folder);

	if (mon) g_object_unref(mon);

	if (config_filename) g_free(config_filename);
}


/**
 * handle signals that are sent to the application
 */
static void signal_handler(int sig)
{
	printf("\n");
	printf("%s", _("Received signal:"));
	printf(" %d (%s)\n", sig, strsignal(sig));

	done_script_error_messages();

	if (sig == SIGINT) {
		exit(EXIT_FAILURE);
	}

}


/**
 *
 */
void print_list(GSList *list)
{
	GSList *temp_list;
	if (list != NULL) {
		temp_list = list;

		while(temp_list) {
			gchar *file_name = temp_list->data;

			if (file_name) {
				if (g_str_has_suffix((gchar*)file_name, ".lua")) {
					printf("%s\n", (gchar*)file_name);
				}
			}
			temp_list = temp_list->next;
		}
	}
}


/**
 *
 */
void print_script_lists()
{
	if (debug)
		printf("------------\n");

	if ((file_window_open_list == NULL) && (file_window_close_list == NULL)) {
		printf("%s", _("No script files found in the script folder - exiting."));
		printf("\n\n");
		exit(EXIT_SUCCESS);
	}

	// If we are running debug mode - print the list of files:
	if (debug) {
		printf(_("List of LUA files handling \"%s\" events in folder:"),
			"window_open");
		printf("\n");
		if (file_window_open_list)
			print_list(file_window_open_list);
	}

	if (debug) {
		printf(_("List of LUA files handling \"%s\" events in folder:"),
			"window_close");
		printf("\n");
		if (file_window_close_list)
			print_list(file_window_close_list);
	}
}


/**
 *
 */
void folder_changed_callback(GFileMonitor *mon,
										GFile *first_file,
										GFile *second_file,
										GFileMonitorEvent event,
										gpointer user_data)
{
	gchar *our_filename = (gchar*)(user_data);

	// If a file is created or deleted, we need to check the file lists again
	if ((event == G_FILE_MONITOR_EVENT_CREATED) ||
		 (event == G_FILE_MONITOR_EVENT_DELETED)) {

		clear_file_lists();

		load_config(our_filename);

		if (debug)
			printf("Files in folder updated!\n - new lists:\n\n");

		print_script_lists();

		if (debug)
			printf("-----------\n");
	}

	// Also monitor if our devilspie2.lua file is changed - since it handles
	// which files are window close or window open scripts.
	if (event == G_FILE_MONITOR_EVENT_CHANGED) {
		if (first_file) {
			gchar *short_filename = g_file_get_basename(first_file);

			if (g_strcmp0(short_filename, "devilspie2.lua")==0) {

				clear_file_lists();

				load_config(our_filename);

				print_script_lists();

				if (debug)
					printf("----------");
			}
		}
	}
}


/**
 * Program main entry
 */
int main(int argc, char *argv[])
{
	static const GOptionEntry options[]={
		{ "debug",			'd',	0,	G_OPTION_ARG_NONE,		&debug,
			N_("Print debug info to stdout")},
		{ "emulate",		'e',	0,	G_OPTION_ARG_NONE,		&emulate,
			N_("Don't apply any rules, only emulate execution")},
		{ "folder",			'f',	0,	G_OPTION_ARG_STRING,		&script_folder,
			N_("Search for scripts in this folder"),N_("FOLDER")},
		{ "version",		'v',	0,	G_OPTION_ARG_NONE,		&show_version,
			N_("Show Devilspie2 version and quit")},
#ifdef HAVE_GTK3
		// libwnck Version Information is only availible if you have
		// libwnck 3.0 or later
		{ "wnck-version",	'w',	0,	G_OPTION_ARG_NONE,		&show_wnck_version,
			N_("Show libwnck version and quit")},
#endif
		{ NULL }
	};

	GError *error = NULL;
	GOptionContext *context;

	// Init gettext stuff
	setlocale(LC_ALL, "");

	bindtextdomain(PACKAGE, LOCALEDIR);
	bind_textdomain_codeset(PACKAGE, "");
	textdomain(PACKAGE);

	gdk_init(&argc, &argv);

	gchar *devilspie2_description =
		g_strdup_printf(_("apply rules on windows"));

	gchar *full_desc_string = g_strdup_printf("- %s", devilspie2_description);

	context=g_option_context_new(full_desc_string);
	g_option_context_add_main_entries(context,options,NULL);
	if (!g_option_context_parse(context, &argc, &argv, &error)) {
		g_print(_("option parsing failed: %s"), error->message);
		printf("\n");
		exit(EXIT_FAILURE);
	}

	g_free(full_desc_string);
	g_free(devilspie2_description);

	// if the folder is NULL, default to ~/.config/devilspie2/
	if (script_folder == NULL) {

		temp_folder = g_build_path(G_DIR_SEPARATOR_S,
										g_get_user_config_dir(),
										"devilspie2",
										NULL);

		// check if the folder does exist
		if (!g_file_test(temp_folder, G_FILE_TEST_IS_DIR)) {

			// - and if it doesn't, create it.
			if (g_mkdir(temp_folder, 0700) != 0) {
				printf("%s", _("Couldn't create the default folder for devilspie2 scripts."));
				printf("\n");
				exit(EXIT_FAILURE);
			}
		}

		script_folder = temp_folder;
	}

	if (show_version) {
		printf("Devilspie2 v%s\n\n", DEVILSPIE2_VERSION);
		exit(EXIT_SUCCESS);
	}
#ifdef HAVE_GTK3
	// libwnck Version Information is only availible if you have
	// libwnck 3.0 or later
	if (show_wnck_version) {
		printf("libwnck v%d.%d.%d\n\n",
		       WNCK_MAJOR_VERSION,
		       WNCK_MINOR_VERSION,
		       WNCK_MICRO_VERSION);
		exit(EXIT_SUCCESS);
	}
#endif

#if (GTK_MAJOR_VERSION >= 3)
	if (!GDK_IS_X11_DISPLAY(gdk_display_get_default())) {
		printf("%s", _("An X11 display is required for devilspie2."));
		printf("\n\n");
		return EXIT_FAILURE;
	}

#endif

	if (init_script_error_messages()!=0) {
		printf("%s", _("Couldn't init script error messages!"));
		printf("\n");
		exit(EXIT_FAILURE);
	}

	config_filename =
		g_build_filename(script_folder, "devilspie2.lua", NULL);

	if (load_config(config_filename)!=0) {

		devilspie_exit();
		return EXIT_FAILURE;
	}

	if (debug) {

		if (emulate) {
			printf("%s", _("Running devilspie2 in debug and emulate mode."));
		} else {
			printf("%s", _("Running devilspie2 in debug mode."));
		}

		printf("\n\n");

		printf(_("Using scripts from folder: %s"), script_folder);

		printf("\n");

		devilspie2_debug = TRUE;
	}

	// Should we only run an emulation (don't modify any windows)
	if (emulate) devilspie2_emulate = emulate;

	GFile *directory_file;
	directory_file = g_file_new_for_path(script_folder);
//	mon = g_file_monitor_directory(directory_file, G_FILE_MONITOR_WATCH_MOUNTS,
	mon = g_file_monitor_directory(directory_file, G_FILE_MONITOR_NONE,
												NULL, NULL);
	if (!mon) {
		printf("%s", _("Couldn't create directory monitor!"));
		printf("\n");
		return EXIT_FAILURE;
	}

	g_signal_connect(mon, "changed", G_CALLBACK(folder_changed_callback),
		(gpointer)(config_filename));

	global_lua_state = init_script();
	print_script_lists();

	if (debug) printf("------------\n");

	// remove stuff cleanly
	atexit(devilspie_exit);

	struct sigaction signal_action;

	sigemptyset(&signal_action.sa_mask);
	signal_action.sa_flags = 0;
	signal_action.sa_handler = signal_handler;

	if (sigaction(SIGINT, &signal_action, NULL) == -1) {
		exit(EXIT_FAILURE);
	}

	init_screens();

	loop=g_main_loop_new(NULL, TRUE);
	g_main_loop_run(loop);

	return EXIT_SUCCESS;
}
